[SYCL] Add support for VLAs in kernel code

premanandrao opened the pull request:
Restriction on using VLAs in sycl kernel code has been removed.
VLAs are represented in the integration header file with a new
kind - SYCLIntegrationHeader::kind_vla.  The information field
encodes the number of dimensions and the size of the underlying
type.  The preceding entries of kind SYCLIntegrationHeader::kind_std_layout
describe the respective extent of each variable dimension.

elizabethandrews commented on clang/test/CodeGenSYCL/kernel-param-vla-ih.cpp:
```diff
@@ -0,0 +1,94 @@
+
+using namespace cl::sycl;
+
+template <typename name, typename Func>
```
Is there a reason you are defining a_kernel instead of using single_task, etc?

premanandrao replied:
No, that is how I started, and I just failed to fix it before submitting.

elizabethandrews commented on clang/test/SemaSYCL/vla-kernel-param.cpp:
```diff
@@ -0,0 +1,132 @@
+// CHECK-NEXT: ImplicitCastExpr {{.*}} 'int *__global *' <LValueToRValue>
+// CHECK-NEXT: DeclRefExpr {{.*}} 'int *__global *' lvalue ParmVar {{.*}} '_arg_' 'int *__global *'
+
+// Check FLA parameters
```
This is a personal preference which you can disregard if you disagree. I prefer having CHECK statements next to what we are testing instead of all together at the end like this. In my opinion, it just makes it easier to read tests. Same comment for CodeGen test above

premanandrao replied:
Thanks, I will make this change.   But before I touch this file, another question.

The contents of SemaSYCL/vla-kernel-param.cpp and CodeGenSYCL/kernel-param-vla-ih.cpp are mostly the same.  The former is used for the AST dump verification and the latter for Integration Header checks.   Should I merge the two into one place?  If so, where?  Is there a preference?

elizabethandrews replied:
I think its better keeping it separate the way you have now.

premanandrao replied:
Please let me know what you think of the rearranged CHECKs.

AaronBallman commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -2230,9 +2216,24 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
-    PointeeTy = SemaRef.getASTContext().getQualifiedType(
-        PointeeTy.getUnqualifiedType(), Quals);
-    QualType ModTy = SemaRef.getASTContext().getPointerType(PointeeTy);
+    // Handle VLAs if this is a pointer to one
```
```suggestion
    // Handle VLAs if this is a pointer to one.
```
Same suggestion to add trailing punctuation applies elsewhere in the file.

premanandrao commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1723,14 +1717,6 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
   }
 
   bool handlePointerType(FieldDecl *FD, QualType FieldTy) final {
```
Thanks, will do.

premanandrao commented on clang/test/SemaSYCL/sycl-restrict.cpp:
```diff
@@ -43,22 +43,18 @@ namespace Check_VLA_Restriction {
 void no_restriction(int p) {
```
One of them is being called (to show that there is no diagnostic generated).   I will remove the other one.

AaronBallman commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -3443,9 +3444,26 @@ class SyclKernelIntHeaderCreator : public SyclKernelFieldHandler {
+    ASTContext &Context = SemaRef.getASTContext();
+    // VLAs are encoded as (Dim << 24) | Size
+    // where Dim is the number of VLA dimensions and
+    // Size is the size of the underlying type
```
Oh, so we don't keep track of what extents are used for each dimension?

premanandrao replied:
Thanks for asking this.   I have redone the comment to indicate that there are as many kind_std_layout entries preceding the vla entry as the number of dimensions; and each of those describe the extent of their respective dimension.

premanandrao commented on clang/test/SemaSYCL/sycl-restrict.cpp:
```diff
@@ -43,22 +43,18 @@ namespace Check_VLA_Restriction {
 void *operator new(std::size_t size, void *ptr) throw() { return ptr; };
 namespace Check_RTTI_Restriction {
 struct A {
+  // expected-note@#call_usage {{called by 'operator()'}}
```
These are not new diagnostics.  If you look at the diagnostics that I removed (e.g. lines 46-55 above) these diagnostics were emitted multiple times (but for different violations).   I just removed the VLA diagnostics and the call-chain notes for it, and just cleaned up the remaining ones to associate it with the place that it is actually generated for.

AaronBallman commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -3443,9 +3444,26 @@ class SyclKernelIntHeaderCreator : public SyclKernelFieldHandler {
-             ((StructDepth) ? SYCLIntegrationHeader::kind_std_layout
-                            : SYCLIntegrationHeader::kind_pointer));
+    QualType PointeeTy = FieldTy->getPointeeType();
+    unsigned int Dim = 0;
```
```suggestion
    unsigned Dim = 0;
```

premanandrao commented on clang/test/SemaSYCL/vla-kernel-param.cpp:
```diff
@@ -0,0 +1,132 @@
+// This test checks that compiler generates correct kernel arguments for
+// VLAs.
+
+template <typename name, typename Func>
```
And the same answer as above.   It was just an oversight.

AaronBallman commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -3443,9 +3444,26 @@ class SyclKernelIntHeaderCreator : public SyclKernelFieldHandler {
+    unsigned Size = Context.getTypeSizeInChars(PointeeTy).getQuantity();
+    if (Dim)
+      Header.addParamDesc(SYCLIntegrationHeader::kind_vla,
+                          static_cast<unsigned>(Size | (Dim << 24)),
```
```suggestion
                          Size | (Dim << 24u),
```

keryell commented:
I have been always a fan of C99 VLA but would not have expected to see them in SYCL. :-)
But it is not clear to me how it is implemented.
Are you just relying on the fact that the size is made from scalars which are captured by the SYCL kernel anyway?
Can you have some tests like
```c++
int vla[x];
// Check size
++x;
// Check size
using array = int[x];
//check size of array
++x;
//check size of array
```
?
I wonder what we can get from code like
```c++
using array = int[&x - &y];
```
since the memory layout of x and y will be different between the host and the device and thus the address difference.

premanandrao commented:
> I have been always a fan of C99 VLA but would not have expected to see them in SYCL. :-)
> But it is not clear to me how it is implemented.
> Are you just relying on the fact that the size is made from scalars which are captured by the SYCL kernel anyway?
> Can you have some tests like
> 
> ```c++
> int vla[x];
> // Check size
> ++x;
> // Check size
> using array = int[x];
> //check size of array
> ++x;
> //check size of array
> ```
> 
> ?

I have modified vla-kernel-param.cpp to add such a case.   Please let me know if that suffices for what you were asking.

Some of what you are asking is regular VLA functionality.   I have not modified those; I presume there are adequate tests for that aspect of VLAs.  What I changed in the test is to show that the right value is being used for the declaration of the VLA.

> I wonder what we can get from code like
> 
> ```c++
> using array = int[&x - &y];
> ```
> 
> since the memory layout of x and y will be different between the host and the device and thus the address difference.

I am not sure what would happen here.   While you are pointing out a scenario with variable array allocation, the same disparity would exist for other cases (such as in an array indexing expression) even currently.

AaronBallman commented:
VLAs can execute code that's typically in an unevaluated context, and there's a part of me that wonder how this interacts with the device/host split and whether there are going to be problems from that. e.g.,
```
int func() {
  int x = 10;
  kernel([&] {
    (void)sizeof(int[x++]); // This operand *is* evaluated!
  });
  return x; // Returns 11, not 10.
}
```

AaronBallman commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -3443,9 +3440,27 @@ class SyclKernelIntHeaderCreator : public SyclKernelFieldHandler {
+    // VLAs are encoded as (Dim << 24) | Size, where Dim is the
+    // number of VLA dimensions and Size is the size of the
+    // underlying type. Dim indicates that an equal number of
+    // preceding entries (of std_layout kind) describe the
```
Is this actually sufficient though? As best I can tell, `std_layout` kinds eventually go through `addParam()`, which only specifies the offset and the size of the dimension, but not what variable it's tracked from. e.g.,
```
int x = 10;
x++;
x += 10;
int vla[x]; // x == 21
```
I don't see how the `x` would be tracked through such that a `sizeof(vla)` on the device side would give the correct answer.

elizabethandrews reviewed (commented):
I think we need proper documentation describing how VLAs are handled in sycl/docs/KernelParameterPassing.md. What do you think? 

VLA processing isn't straightforward and having step-by-step explanation (similar to how array handling is explained in this document) might be helpful. I don't fully understand how the arguments and initializers are generated for VLA right now

Can you also check what happens with opt-report for VLAs?

keryell commented:
By the way, what is the motivating code or application behind this extension ? Because this seems to represent some work, not even talking about the philosophical problems from a C++ point of view (`sizeof` is no longer a `constexpr` value).
